charts/partials/templates/_affinity.tpl

```diff
@@ -1,22 +1,44 @@
 {{ define "linkerd.pod-affinity" -}}
-affinity:
-  podAntiAffinity:
-    preferredDuringSchedulingIgnoredDuringExecution:
-    - podAffinityTerm:
-        labelSelector:
-          matchExpressions:
-          - key: {{ default "linkerd.io/control-plane-component" .label }}
-            operator: In
-            values:
-            - {{ .component }}
-        topologyKey: failure-domain.beta.kubernetes.io/zone
-      weight: 100
-    requiredDuringSchedulingIgnoredDuringExecution:
-    - labelSelector:
+{{- if not .Values.enablePodAntiAffinity -}}
+podAntiAffinity: null
+{{- else -}}
+podAntiAffinity:
+  preferredDuringSchedulingIgnoredDuringExecution:
+  - podAffinityTerm:
+      labelSelector:
         matchExpressions:
         - key: {{ default "linkerd.io/control-plane-component" .label }}
           operator: In
           values:
           - {{ .component }}
-      topologyKey: kubernetes.io/hostname
+      topologyKey: failure-domain.beta.kubernetes.io/zone
+    weight: 100
+  requiredDuringSchedulingIgnoredDuringExecution:
+  - labelSelector:
+      matchExpressions:
+      - key: {{ default "linkerd.io/control-plane-component" .label }}
+        operator: In
+        values:
+        - {{ .component }}
+    topologyKey: kubernetes.io/hostname
+{{- end }}
+{{- end }}
+
+{{ define "linkerd.node-affinity" -}}
+{{- if not .Values.nodeAffinity -}}
+nodeAffinity: null
+{{- else -}}
+nodeAffinity:
+{{- toYaml .Values.nodeAffinity | trim | nindent 2 }}
+{{- end }}
+{{- end }}
+
+{{ define "linkerd.affinity" -}}
+{{- if not .Values.enablePodAntiAffinity -}}
+affinity: null
+{{- else -}}
+affinity:
+{{- include "linkerd.pod-affinity" . | nindent 2 }}
+{{- include "linkerd.node-affinity" . | nindent 2 }}
+{{- end }}
```

Comment:
I think it might be cleaner to express all the conditional logic here and remove it from the `linkerd.pod-affinity` and `linkerd.node-affinity` templates
```suggestion
{{- if or .Values.enablePodAntiAffinity or .Values.nodeAffinity -}}
affinity:
{{- if .Values.enablePodAntiAffinity -}}
{{- include "linkerd.pod-affinity" . | nindent 2 }}
{{- end }}
{{- if .Values.nodeAffinity -}}
{{- include "linkerd.node-affinity" . | nindent 2 }}
{{- end }}
{{- end}}
```

Reply:
Totally agree.

Reply:
Agree